File: 01_language_comparison/math_and_miscellaneous/prime.scala
object PrimeChecker {
  // Checks whether a number is a prime
  def isPrime(n: Int): Boolean = {
    if (n <= 1) return false
    if (n == 2) return true
    if (n % 2 == 0) return false

    val sqrtN = math.sqrt(n).toInt
    // Only check odd divisors up to sqrt(n)
    !(3 to sqrtN by 2).exists(n % _ == 0)
  }

  def main(args: Array[String]): Unit = {
    println(isPrime(31)) // Example usage
  }
}

File: 03_error_handling_and_debugging/math_and_miscellaneous/prime.rb
# Checks whether a number is a prime
def prime?(n)
    return false if n <= 1
    return true if n == 2
    return false if n.even?
  
    # Only check odd divisors up to the square root of n
    max = Math.sqrt(n).to_i
    (3..max).step(2).each do |i|
      return false if n % i == 0
    end
    true
  end
  